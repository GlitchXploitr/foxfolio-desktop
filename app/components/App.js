// @flow
import type { Children } from 'react';
import React, { Component } from 'react';
import { withStyles } from 'material-ui/styles';
import Drawer from 'material-ui/Drawer';
import { AppBar, List, ListItem, ListItemIcon, ListItemText, Toolbar, Typography } from 'material-ui';
import { Cloud, Dashboard } from 'material-ui-icons';
import { NavLink } from 'react-router-dom';
import { continuouslyFetchTransactions } from '../actions/transactions';

const drawerWidth = 250;

const styles = theme => ({
  appFrame: {
    width: '100%',
    height: '100%',
  },
  content: {
    backgroundColor: theme.palette.background.default,
    padding: theme.spacing.unit * 3,
    height: 'calc(100% - 56px)',
    marginLeft: drawerWidth,
  },
  drawerPaper: {
    width: drawerWidth,
  },
  bottomList: {
    marginTop: 'auto',
    flexGrow: 0,
  },
});

class App extends Component {
  props: {
    dispatch: (action: any) => void,
    children: Children,
    classes: any
  };

  componentDidMount() {
    this.props.dispatch(continuouslyFetchTransactions());
  }

  render() {
    const { classes, children } = this.props;
    return (
      <div className={classes.appFrame}>
        <Drawer type="permanent" classes={{ paper: classes.drawerPaper }}>
          <AppBar position="static" color="primary" elevation={1}>
            <Toolbar>
              <Typography type="title" color="inherit">
                Awesome Portfolio
              </Typography>
            </Toolbar>
          </AppBar>
          <List>
            <ListItem button component={NavLink} exact to="/">
              <ListItemIcon>
                <Dashboard/>
              </ListItemIcon>
              <ListItemText primary="Portfolio"/>
            </ListItem>
          </List>
          <List className={classes.bottomList}>
            <ListItem button component={NavLink} to="/sources">
              <ListItemIcon>
                <Cloud/>
              </ListItemIcon>
              <ListItemText primary="Sources"/>
            </ListItem>
          </List>
        </Drawer>
        <main className={classes.content}>
          {children}
        </main>
      </div>
    );
  }
}

export default withStyles(styles)(App);
