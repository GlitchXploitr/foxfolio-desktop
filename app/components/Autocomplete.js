// @flow
import React from 'react';
import Downshift from 'downshift';
import { MenuItem, Paper, TextField, withStyles } from 'material-ui';

const styles = {};

type Props = {
  id: string,
  label: string,
  classes: Object,
  onChange: (string) => void,
  value: string,
  items: string[]
};

const isSelected = (selection: string, item: string) =>
  !selection || item.toLowerCase().includes(selection.toLowerCase());

function UnstyledAutocomplete(props: Props) {
  const { classes, id, label, items, onChange, value } = props;

  return (
    <Downshift onChange={onChange} defaultSelectedItem={value}>
      {({ getInputProps, getItemProps, isOpen, inputValue, highlightedIndex }) => (
        <div>
          <TextField
            fullWidth
            autoFocus
            label={label}
            InputProps={getInputProps({ id })}
          />
          {isOpen ? (
            <Paper square className={classes.paper}>
              {items
                .filter(exchange => isSelected(inputValue, exchange))
                .slice(0, 5)
                .map((suggestion, index) => (
                  <MenuItem
                    {...getItemProps({ item: suggestion })}
                    key={suggestion}
                    selected={highlightedIndex === index}
                    component="div"
                  >
                    {suggestion}
                  </MenuItem>),
                )}
            </Paper>
          ) : null}
        </div>
      )}
    </Downshift>
  );
}

export const Autocomplete = withStyles(styles)(UnstyledAutocomplete);
