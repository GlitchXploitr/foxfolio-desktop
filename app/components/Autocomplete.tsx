import Downshift from 'downshift';
import { MenuItem, Paper, TextField, withStyles } from 'material-ui';
import { StyleRules } from 'material-ui/styles';
import React from 'react';

const styles: StyleRules = {
  container: { height: 50, position: 'relative' },
  paper: { position: 'absolute', width: '100%', zIndex: 9999 },
};

interface Props {
  id: string;
  label: string;
  onChange: (item: string) => any;
  value: string;
  items: string[];
}

const isSelected = (selection: string | null, item: string) =>
  !selection || item.toLowerCase().includes(selection.toLowerCase());

export const Autocomplete = withStyles(styles)<Props>(
  ({ classes, id, label, items, onChange, value }) => (
    <Downshift
      onChange={onChange}
      itemToString={i => (i == null ? '' : String(i))}
      defaultSelectedItem={value}
    >
      {({ getInputProps, getItemProps, isOpen, inputValue, highlightedIndex }) => (
        <div className={classes.container}>
          <TextField fullWidth autoFocus label={label} InputProps={getInputProps({ id })} />
          {isOpen ? (
            <Paper square className={classes.paper}>
              {items
                .filter(exchange => isSelected(inputValue, exchange))
                .slice(0, 5)
                .map((suggestion, index) => (
                  <MenuItem
                    {...getItemProps({ item: suggestion })}
                    key={suggestion}
                    selected={highlightedIndex === index}
                    component="div"
                  >
                    {suggestion}
                  </MenuItem>
                ))}
            </Paper>
          ) : null}
        </div>
      )}
    </Downshift>
  )
);
