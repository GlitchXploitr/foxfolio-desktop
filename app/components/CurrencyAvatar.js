// @flow
import React from 'react';
import { Avatar } from 'material-ui';
import { AccountBalance, HelpOutline } from 'material-ui-icons';
import type { Coinlist } from '../reducers/coinlist/types.d';
import { getFiatCurrencies } from '../utils/fiatCurrencies';

type Props = {
  asset: string,
  coinlist: Coinlist,
  fiatClass: Object
};

export const CurrencyAvatar = ({ asset, coinlist, fiatClass }: Props) => {
  let avatar = <Avatar><HelpOutline/></Avatar>;
  if (coinlist[asset]) {
    avatar = <Avatar src={`https://www.cryptocompare.com${coinlist[asset].ImageUrl}`}/>;
  } else if (getFiatCurrencies().includes(asset)) {
    avatar = <Avatar className={fiatClass}><AccountBalance/></Avatar>;
  }
  return avatar;
};
