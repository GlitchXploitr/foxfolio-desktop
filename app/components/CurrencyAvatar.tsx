import { Avatar } from 'material-ui';
import { AccountBalance, HelpOutline } from 'material-ui-icons';
import React from 'react';

import { Coinlist } from '../reducers/coinlist';
import { getFiatCurrencies } from '../utils/fiatCurrencies';

interface Props {
  asset: string;
  coinlist: Coinlist;
  fiatClass: string;
}

export const CurrencyAvatar = ({ asset, coinlist, fiatClass }: Props) => {
  let avatar = (
    <Avatar>
      <HelpOutline />
    </Avatar>
  );
  if (coinlist[asset]) {
    avatar = <Avatar src={`https://www.cryptocompare.com${coinlist[asset].ImageUrl}`} />;
  } else if (getFiatCurrencies().includes(asset)) {
    avatar = (
      <Avatar className={fiatClass}>
        <AccountBalance />
      </Avatar>
    );
  }
  return avatar;
};
