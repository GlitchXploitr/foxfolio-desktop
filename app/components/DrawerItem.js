// @flow
import type { Node } from 'react';
import React from 'react';
import { ListItem, ListItemIcon, ListItemText } from 'material-ui';
import { NavLink } from 'react-router-dom';

type Props = {
  icon: Node,
  label: string,
  route: string
};

export function DrawerItem({ icon, label, route }: Props) {
  return (
    <ListItem button component={NavLink} exact to={route}>
      <ListItemIcon>
        {icon}
      </ListItemIcon>
      <ListItemText primary={label}/>
    </ListItem>
  );
}
