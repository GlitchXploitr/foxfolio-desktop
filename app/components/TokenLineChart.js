// @flow
import React, { Component } from 'react';
import type { MapStateToProps } from 'react-redux';
import { connect } from 'react-redux';
import moment from 'moment';
import { Area, AreaChart, CartesianGrid, ResponsiveContainer, Tooltip, XAxis, YAxis } from 'recharts';
import { bindActionCreators } from 'redux';
import { LinearProgress, withTheme } from 'material-ui';
import * as TickerActions from '../actions/ticker';

type Props = {
  fsym: string,
  tsym: string,
  requestHistory: (fsym: string, tsym: string) => void,
  history: Object,
  theme: Object
};

export class RawTokenLineChart extends Component<Props> {
  componentWillMount() {
    const { requestHistory, fsym, tsym } = this.props;
    requestHistory(fsym, tsym);
  }

  render() {
    const { history, fsym, tsym, theme } = this.props;

    if (fsym !== tsym && history[fsym] && history[fsym][tsym]) {
      const reducedHistory = history[fsym][tsym].data;
      return (
        <ResponsiveContainer height={300}>
          <AreaChart data={reducedHistory} margin={{ left: 40, top: 5, bottom: 5, right: 5 }}>
            <Area
              stroke={theme.palette.text.secondary}
              fill={theme.palette.text.divider}
              dataKey="close"
              isAnimationActive={false}
            />
            <XAxis
              dataKey="time"
              tick={{ fill: theme.palette.text.secondary, strokeWidth: 0 }}
              tickFormatter={tick => moment.unix(tick).format('HH:mm')}
            />
            <YAxis
              interval="preserveStartEnd"
              tick={{ fill: theme.palette.text.secondary, strokeWidth: 0 }}
              tickFormatter={tick => tick.toPrecision(6)}
              domain={[dataMin => dataMin - (0.02 * dataMin), dataMax => dataMax + (0.02 * dataMax)]}
            />
            <CartesianGrid stroke={theme.palette.text.divider} strokeDasharray="3 3"/>
            <Tooltip
              itemStyle={{ color: theme.palette.text.secondary }}
              wrapperStyle={{ backgroundColor: theme.palette.background.paper }}
              labelFormatter={() => ''}
            />
          </AreaChart>
        </ResponsiveContainer>
      );
    } else if (fsym !== tsym) {
      return <LinearProgress/>;
    }
    return <div>No chart for currency pair {fsym}-{tsym}</div>;
  }
}

const mapStateToProps: MapStateToProps<*, *, *> = state => ({
  history: state.ticker.history,
});
const mapActionsToProps = (dispatch: Dispatch) => bindActionCreators(TickerActions, dispatch);

export const TokenLineChart = connect(mapStateToProps, mapActionsToProps)(withTheme()(RawTokenLineChart));
