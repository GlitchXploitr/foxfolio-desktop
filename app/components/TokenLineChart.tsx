import { LinearProgress, withTheme } from 'material-ui';
import { Theme } from 'material-ui/styles';
import { WithTheme } from 'material-ui/styles/withTheme';
import moment from 'moment';
import React, { Component } from 'react';
import { connect, Dispatch, MapStateToProps } from 'react-redux';
import {
  Area,
  AreaChart,
  CartesianGrid,
  ResponsiveContainer,
  Tooltip,
  XAxis,
  YAxis,
} from 'recharts';
import { bindActionCreators } from 'redux';

import * as TickerActions from '../actions/ticker';
import { GlobalState } from '../reducers';
import { History } from '../reducers/ticker';

interface StateProps {
  history: History;
}

interface DispatchProps {
  requestHistory: (fsym: string, tsym: string) => void;
}

interface OwnProps {
  fsym: string;
  tsym: string;
}

export const ThemedTokenLineChart = withTheme()(
  class extends Component<StateProps & DispatchProps & OwnProps & WithTheme> {
    public componentWillMount() {
      const { requestHistory, fsym, tsym } = this.props;
      requestHistory(fsym, tsym);
    }

    public render() {
      const { history, fsym, tsym, theme } = this.props;

      if (fsym !== tsym && history[fsym] && history[fsym][tsym]) {
        const reducedHistory = history[fsym][tsym].data;
        return (
          <ResponsiveContainer height={300}>
            <AreaChart data={reducedHistory} margin={{ left: 40, top: 5, bottom: 5, right: 5 }}>
              <Area
                stroke={theme.palette.text.secondary}
                fill={theme.palette.text.divider}
                dataKey="close"
                isAnimationActive={false}
              />
              <XAxis
                dataKey="time"
                tick={{ fill: theme.palette.text.secondary, strokeWidth: 0 }}
                tickFormatter={tick => moment.unix(tick).format('HH:mm')}
              />
              <YAxis
                interval="preserveStartEnd"
                tick={{ fill: theme.palette.text.secondary, strokeWidth: 0 }}
                tickFormatter={tick => tick.toPrecision(6)}
                domain={[dataMin => dataMin - 0.02 * dataMin, dataMax => dataMax + 0.02 * dataMax]}
              />
              <CartesianGrid stroke={theme.palette.text.divider} strokeDasharray="3 3" />
              <Tooltip
                itemStyle={{ color: theme.palette.text.secondary }}
                wrapperStyle={{ backgroundColor: theme.palette.background.paper }}
                labelFormatter={() => ''}
              />
            </AreaChart>
          </ResponsiveContainer>
        );
      } else if (fsym !== tsym) {
        return <LinearProgress />;
      }
      return (
        <div>
          No chart for currency pair {fsym}-{tsym}
        </div>
      );
    }
  }
);

const mapStateToProps: MapStateToProps<StateProps, OwnProps, GlobalState> = state => ({
  history: state.ticker.history,
});
const mapActionsToProps = (dispatch: Dispatch<GlobalState>) =>
  bindActionCreators(TickerActions, dispatch);

export const TokenLineChart = connect<StateProps, DispatchProps, OwnProps, GlobalState>(
  mapStateToProps,
  mapActionsToProps
)(ThemedTokenLineChart);
