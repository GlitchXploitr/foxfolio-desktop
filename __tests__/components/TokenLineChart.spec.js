// @flow
import { createMount } from 'material-ui/test-utils';
import * as React from 'react';
import { RawTokenLineChart } from '../../app/components/TokenLineChart';

let mount;

beforeAll(() => {
  mount = createMount();
});

test('Render token line chart', () => {
  const fsym = 'ETH';
  const tsym = 'USD';

  // eslint-disable-next-line no-unused-vars
  const requestHistory = (a, b) => {
  };
  const history = { ETH: { USD: { data: [{ close: 10, time: 1000000 }] } } };
  const theme = { palette: { text: { secondary: '', divider: '' }, background: { paper: '' } } };

  const wrapper = mount(
    <RawTokenLineChart
      fsym={fsym}
      tsym={tsym}
      requestHistory={requestHistory}
      history={history}
      theme={theme}
    />);
  expect(wrapper).toMatchSnapshot();
});
