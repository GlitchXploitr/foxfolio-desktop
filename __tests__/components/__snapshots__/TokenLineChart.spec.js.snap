// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render token line chart 1`] = `
<RawTokenLineChart
  fsym="ETH"
  history={
    Object {
      "ETH": Object {
        "USD": Object {
          "data": Array [
            Object {
              "close": 10,
              "time": 1000000,
            },
          ],
        },
      },
    }
  }
  requestHistory={[Function]}
  theme={
    Object {
      "palette": Object {
        "background": Object {
          "paper": "",
        },
        "text": Object {
          "divider": "",
          "secondary": "",
        },
      },
    }
  }
  tsym="USD"
>
  <ResponsiveContainer
    debounce={0}
    height={300}
    width="100%"
  >
    <div
      className="recharts-responsive-container"
      style={
        Object {
          "height": 300,
          "maxHeight": undefined,
          "minHeight": undefined,
          "minWidth": undefined,
          "width": "100%",
        }
      }
    >
      <AreaChart
        barCategoryGap="10%"
        barGap={4}
        data={
          Array [
            Object {
              "close": 10,
              "time": 1000000,
            },
          ]
        }
        height={300}
        layout="horizontal"
        margin={
          Object {
            "bottom": 5,
            "left": 40,
            "right": 5,
            "top": 5,
          }
        }
        reverseStackOrder={false}
        stackOffset="none"
        width={0}
      />
      <ResizeDetector
        handleHeight={true}
        handleWidth={true}
        onResize={[Function]}
      >
        <div
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "overflow": "hidden",
              "position": "absolute",
              "right": 0,
              "top": 0,
              "visibility": "hidden",
              "zIndex": -1,
            }
          }
        >
          <div
            onScroll={[Function]}
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "overflow": "hidden",
                "position": "absolute",
                "right": 0,
                "top": 0,
                "visibility": "hidden",
                "zIndex": -1,
              }
            }
          >
            <div
              style={
                Object {
                  "height": 10,
                  "left": 0,
                  "position": "absolute",
                  "top": 0,
                  "width": 10,
                }
              }
            />
          </div>
          <div
            onScroll={[Function]}
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "overflow": "hidden",
                "position": "absolute",
                "right": 0,
                "top": 0,
                "visibility": "hidden",
                "zIndex": -1,
              }
            }
          >
            <div
              style={
                Object {
                  "height": "200%",
                  "left": 0,
                  "position": "absolute",
                  "top": 0,
                  "width": "200%",
                }
              }
            />
          </div>
        </div>
      </ResizeDetector>
    </div>
  </ResponsiveContainer>
</RawTokenLineChart>
`;
